### PR TITLE
[processor/redaction] add support for redacting metrics and logs attributes

### PR DESCRIPTION
**Pull request by bacherfl**
**Description:** This PR extends the redaction processor to also support the redaction of attributes within logs and metrics.

**Link to tracking Issue:** #34479 

**Testing:** Extended the existing unit tests to also cover the redaction of logs and metrics

**Documentation:** Adapted the readme to reflect the changes

### CODE REVIEW

File: processor/redactionprocessor/processor.go [hunk 1]
```diff
@@ -79,6 +97,37 @@ func (s *redaction) processResourceSpan(ctx context.Context, rs ptrace.ResourceS
...
+	rsAttrs := rm.Resource().Attributes()
+
+	s.processAttrs(ctx, rsAttrs)
+
+	for j := 0; j < rm.ScopeMetrics().Len(); j++ {
+		ils := rm.ScopeMetrics().At(j)
+		s.processAttrs(ctx, ils.Scope().Attributes())
+		for k := 0; k < ils.Metrics().Len(); k++ {
+			metric := ils.Metrics().At(k)
+			s.processAttrs(ctx, metric.Metadata())
```
Comment (bacherfl) [inline processor/redactionprocessor/processor.go:126]: there is no `Attributes` property within the metrics or its data points, but `Metadata` seems to have a similar meaning, so passing on those to the `processAttrs` function

Comment (mx-psi) [reply processor/redactionprocessor/processor.go:126]: See https://github.com/open-telemetry/opentelemetry-proto/blob/e6923b74ccb43c884308b589cc00f972c12b4055/opentelemetry/proto/metrics/v1/metrics.proto#L212-L213, these are not the attributes you are looking for.

You can take inspiration from https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/eba20f5bedcd1ceeb43348d2c8e7ca524c0ecc77/processor/attributesprocessor/attributes_metric.go#L64 on how to do this

**Review by mx-psi — DISMISSED** [review]

File: processor/redactionprocessor/processor.go (hunk 1 above)
Comment (bacherfl) [reply processor/redactionprocessor/processor.go:126]: thanks for the link - I will adapt this accordingly then

Comment (bacherfl) [reply processor/redactionprocessor/processor.go]: @mx-psi this should be fixed now

File: processor/redactionprocessor/README.md [hunk 2]
```diff
@@ -3,19 +3,21 @@
...
 
+[alpha]: https://github.com/open-telemetry/opentelemetry-collector#alpha
 [beta]: https://github.com/open-telemetry/opentelemetry-collector#beta
 [contrib]: https://github.com/open-telemetry/opentelemetry-collector-releases/tree/main/distributions/otelcol-contrib
 <!-- end autogenerated section -->
 
-This processor deletes span attributes that don't match a list of allowed span
-attributes. It also masks span attribute values that match a blocked value
-list. Span attributes that aren't on the allowed list are removed before any
+This processor deletes span/log/metric attributes that don't match a list of allowed
```
Comment (TylerHelmuth) [inline processor/redactionprocessor/README.md:18]: metrics don't have attributes, lets use datapoints whenever we're referencing attributes.

File: processor/redactionprocessor/processor.go [hunk 3]
```diff
@@ -79,6 +97,64 @@ func (s *redaction) processResourceSpan(ctx context.Context, rs ptrace.ResourceS
...
+}
+
+func (s *redaction) processResourceMetric(ctx context.Context, rm pmetric.ResourceMetrics) {
+	rsAttrs := rm.Resource().Attributes()
+
+	s.processAttrs(ctx, rsAttrs)
+
+	for j := 0; j < rm.ScopeMetrics().Len(); j++ {
+		ils := rm.ScopeMetrics().At(j)
+		s.processAttrs(ctx, ils.Scope().Attributes())
```
Comment (TylerHelmuth) [inline processor/redactionprocessor/processor.go:123]: We process scope attributes here, but not for logs or traces

Comment (bacherfl) [reply processor/redactionprocessor/processor.go:123]: good catch - as the scope attributes haven't been processed for traces before, should I add the processing of scope attributes to traces and logs, or remove that from here?

Comment (bacherfl) [reply processor/redactionprocessor/processor.go]: @TylerHelmuth I removed the processing of scope attributes for now, as these were also not handled by the span processor originally

**Review by TylerHelmuth — APPROVED** [review]

**Review by mx-psi — APPROVED** [review]

**Pull request merged**